src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -136,66 +109,106 @@ private enum BatchWriterState
             Error
         }
 
+        /// <summary>
+        /// >The batch-specific URL converter that stores the content IDs found in a changeset and supports resolving cross-referencing URLs.
+        /// </summary>
+        internal ODataBatchPayloadUriConverter PayloadUriConverter

```

Comment:
>internal [](start = 8, length = 8)

Ideally, all access to the PayloadUriConverter would be in the base class, so we wouldn't need to expose this to the derived class.  Otherwise we would have to make this protected (not internal) since it shouldn't matter whether the derived class is in this assembly or a different assembly.  However, that would require making the ODataBatchPayloadUriConverter public, which I'd rather not do...

Reply:
This field cannot be encapsulated in base class because it needs to be used in CreateOperationRequestMessageImplementation, WriteEndChangesetImplementation, etc that are belonged to derived classes.
So we are OK to keep it as internal, correct? Resolved it now, otherwise please re-open.

---
In reply to: [141140579](https://github.com/OData/odata.net/pull/960#discussion_r141140579) [](ancestors = 141140579)

Reply:
When we refactor ReadImplementation to the base class, it can verfiy that the PayloadUriConvert does not already contain the ContentId before adding it, so we don't need to reference if from the CreateOperationResponseMessage in the derived class.

---
In reply to: [142291342](https://github.com/OData/odata.net/pull/960#discussion_r142291342) [](ancestors = 142291342,141140579)

Reply:
It is also used for CreateOperation[Request|Response]MessageImplementation methods. To support the private encapsulation in base class, the following wrappers (protected) are added:
BuildOperationRequestMeassage
BuildOperationResponseMessage
CreateOperationRequestUriWrapper
AddToPayloadUriConverter
RememberContentIdHeader

---
In reply to: [142823581](https://github.com/OData/odata.net/pull/960#discussion_r142823581) [](ancestors = 142823581,142291342,141140579)

Reply:
See separate comments.  There is no reason for most of this logic to be in the derived class.  The base class can take care of remembering the contentId, wapping the payload, and adding the id tot he payload converter.

---
In reply to: [143319559](https://github.com/OData/odata.net/pull/960#discussion_r143319559) [](ancestors = 143319559,142823581,142291342,141140579)